rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,270 @@
+/* EcDh functions. For further documentation please refer to symcrypt.h */
+
+use crate::curve_type::*;
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+/// EcDhKey is a wrapper around symcrypt_sys::PSYMCRYPT_ECKEY, this is to let rust handle the free'ing of this pointer
+/// EcDhKey must be allocated after EcDhExpandedCurve and free'd before EcDhExpandedCurve is free'd
+struct EcDhKey {
+    inner: symcrypt_sys::PSYMCRYPT_ECKEY,
+    expanded_curve: EcDhExpandedCurve,
+}
+
+struct EcDhExpandedCurve(symcrypt_sys::PSYMCRYPT_ECURVE);
+
+impl Drop for EcDhExpandedCurve {
+    fn drop(&mut self) {
+        unsafe { symcrypt_sys::SymCryptEcurveFree(self.0) }
+    }
+}
+
+impl EcDhKey {
+    pub fn new(curve: CurveType) -> Result<Self, SymCryptError> {
+        unsafe {
+            // Allocating expanded_curve needed for key derivation. This pointer must be dropped after the key is dropped.
+            let curve_ptr = symcrypt_sys::SymCryptEcurveAllocate(convert_curve(curve), 0);
+            if curve_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+            // expanded_curve needs to be wrapped to properly free the curve in the case there is an error in EcDhKey initialization
+            let expanded_curve = EcDhExpandedCurve(curve_ptr);
```

Comment:
I'd factor the `SymCryptEcurveAllocate` code into `EcDhExpandedCurve::new` and then replace the last line here with `let expanded_curve = EcDhExpandedCurve::new()?;` but it doesn't really matter since these are all internal implementation details.

Reply:
+1